src/harness/reference_models/geo/utils.py

```diff
@@ -42,6 +42,48 @@ def GeoJsonToShapelyGeometry(geo):
   return geo
 
 
+def GridPolygon(poly, res_arcsec):
+  """Grids a polygon or multi-polygon.
+
+  This performs regular gridding of a polygon in PlateCarree (equirectangular)
+  projection (ie with fixed step in degrees in lat/lon space).
+  Points falling in the boundary of polygon will be included.
+
+  Args:
+    poly: A Polygon or MultiPolygon in WGS84 or NAD83, defined either
+      as a `shapely` or GeoJSON dict geometries.
+    res_arcsec: The resolution (in arcsec) used for regular gridding.
+
+  Returns:
+    A list of (lon, lat) defining the grid points.
+  """
+  if isinstance(poly, dict):
+    poly = GeoJsonToShapelyGeometry(poly)
+    return GridPolygon(poly, res_arcsec)
+
+  if isinstance(poly, sgeo.MultiPolygon):
+    # Optional block: for MultiPolygons, we process per polygon
+    # to avoid inefficiencies if polygons largely disjoint.
+    pts = ops.unary_union(
+        [sgeo.asMultiPoint(GridPolygon(p, res_arcsec))
+         for p in poly])
+    return [(p.x, p.y) for p in pts]
+
+  res = res_arcsec / 3600.
+  bounds = poly.bounds
+  lng_min = int(bounds[0] / res) * res
+  lat_min = int(bounds[1] / res) * res
+  lng_max = int(bounds[2] / res + 1) * res
+  lat_max = int(bounds[3] / res + 1) * res
+  mesh_lng, mesh_lat = np.mgrid[lng_min:lng_max:res,
+                                lat_min:lat_max:res]
+  points = np.vstack((mesh_lng.ravel(), mesh_lat.ravel())).T
+  pts = poly.intersection(sgeo.asMultiPoint(points))
+  if isinstance(pts, sgeo.Point):
+    return [(pts.x, pts.y)]
+  return [(p.x, p.y) for p in pts]
```

Comment:
what gets returned if isinstance(pts, sgeo.Point) is False?

Reply:
FYI, I have done a further review cycle on the PAT.
If Point is false, then it is a MultiPoint and it returns a list of (lon,lat) tuples.